### PR TITLE
Introduce a class to handle app configuration

### PR DESCRIPTION
**Pull request by ashmaroli**
Summary:
  - initialize new class `WAB::Configuration`
  - parse UNIX-style config files (`*.conf`) into a nested Ruby Hash object
  - add foundation to support runner configuration in JSON or YAML format
  - start `Impl::Shell` instance supplied with configuration object from the new class
  - other minor associated changes

### CODE REVIEW

File: bin/wabur [hunk 1]
```diff
@@ -12,26 +12,25 @@ require 'logger'
...
 verbose_cnt = 0
 
 opts = OptionParser.new(%{Usage: wabur [options]
 
-A WAB Runner that is configured with a unix style conf file.
+A WAB Runner that is configured with a UNIX style config file.
 })
-opts.on('-v', 'verbose')                                     { verbose_cnt += 1 }
-opts.on('-c', '--conf String', String, 'configuration file') { |c| cfg_file = c }
+opts.on('-s', '--source PATH', String, 'Source directory')   { |d| options['source'] = d }
```
Comment (ohler55) [inline bin/wabur]: Sticking with the Ruby and irb conventions `-I` should be used to add to the load path. That should be allowed multiple times as well.

File: lib/wab.rb [hunk 2]
```diff
@@ -4,6 +4,7 @@ module WAB
 end
 
 require 'wab/controller'
+require 'wab/configuration'
```
Comment (ohler55) [inline lib/wab.rb]: This should not be part of the WAB module but in WAB::Impl. It is not required anywhere else. With the exception of the default Controller behavior all the other WAB module classes are duck-types or used to define the API like errors.

File: lib/wab/impl/model.rb [hunk 3]
```diff
@@ -21,7 +22,7 @@ def initialize(dir)
         @cnt = 0
         @map = {}
         @lock = Thread::Mutex.new
-        Dir.mkdir(@dir) unless @dir.nil? || Dir.exist?(@dir)
+        FileUtils.mkdir_p(@dir) unless @dir.nil? || Dir.exist?(@dir)
```
Comment (ohler55) [inline lib/wab/impl/model.rb]: Much better.

**Comment by ohler55** [comment]
Looking good. Just a couple of points.

File: bin/wabur (hunk 1 above)
Comment (ashmaroli) [reply bin/wabur]: I'm not improving the executable entirely in this PR..
Regarding the switch, I opted for `--source` to denote the directory path from where the app will be run.. In the `script/start-sample`, this is handled by the shell variable `$sampledir`

So what's your final suggestion..? I'm getting a sort of mixed vibe here..

File: lib/wab.rb (hunk 2 above)
Comment (ashmaroli) [reply lib/wab.rb]: That makes sense.. the config is ultimately for `WAB::Impl::Shell`..

File: lib/wab/impl/model.rb (hunk 3 above)
Comment (ashmaroli) [reply lib/wab/impl/model.rb]: For some reason, `Dir.mkdir` chokes when I try to run the tests locally.. so there's every chance Travis won't run properly without this change..

File: bin/wabur (hunk 1 above)
Comment (ohler55) [reply bin/wabur]: I guess it got a bit confusing. For now just a rename would be great. 'source' is easily confused with 'source code' which is not what is meant here.

File: lib/wab/impl/model.rb (hunk 3 above)
Comment (ohler55) [reply lib/wab/impl/model.rb]: The key there is mkdir will not make the parent directories if they are not there and mkdir_p will.

Comment (ashmaroli) [reply lib/wab/impl/model.rb]: I agree.. but it works fine on the `develolp` branch..

Comment (ohler55) [reply lib/wab/impl/model.rb]: Maybe the parent directory was already there.

**Pull request merged**